### PR TITLE
--resource-version flag for kubectl GETs and LISTs

### LINKED ISSUE #965: Support for setting resourceVersion in kubectl get
**What would you like to be added**:

Support for `--resource-version=X` for `kubectl get`
(or maybe or simpler flag to just force a read from cache by setting RV=0, such as --cache ?) 

**Why is this needed**:
Today, kubectl get only supporting LISTing with RV="" which on large clusters can be slow for users and impact the control plane and etcd in particular.

I did a few tests on large cluster with curl and the difference can be very significant:

```
time curl  -H "Accept: application/json;as=Table;v=v1beta1;g=meta.k8s.io, application/json" -H "Authorization: Bearer X" 'https://cluster.dog/api/v1/pods?labelSelector=app=A'
real	0m3.658s

time curl -H "Accept: application/json;as=Table;v=v1beta1;g=meta.k8s.io, application/json" -H "Authorization: Bearer X" 'https://cluster.dog/api/v1/pods?labelSelector=app=A&resourceVersion=0'
real	0m0.079s
```
Of course this is an extreme example:
- test on one of our large clusters (30k pods)
- queries not scoped to a namespace
- result of filtering is only a few pods but the query hitting etcd still needs to retrieve all pods before filtering on apiserver

I'd be more than happy to provide a PR but I'm not familiar with the codebase so any (even small) guidance would be appreciated. Here is what I think so far but it's very likely I am missing/misunderstanding more than a few things:
- add resourceVersion to the Builder struct
- add ResourceVersionParam to Builder()
- add a new RetrieveByVersion in addition to RetrieveLatest to Info 

cc @wojtek-t because we discussed this on slack earlier this week

### PR DESCRIPTION
**Pull request by aofekiko**
#### What type of PR is this?

/kind feature

#### What this PR does / why we need it:

Implements the `--resource-version` flag which allows setting the ResourceVersion parameter on GET and LIST requests with kubectl which will allow users to use the different [semantics for get and list](https://kubernetes.io/docs/reference/using-api/api-concepts/#semantics-for-get-and-list) that rely on ResourceVersion.

As stated in the related issue:
> "Today, kubectl get only supports LISTing with ResourceVersion="" which on large clusters can be slow for users and impact the control plane and etcd in particular."

The ability to retrieve stale lists of objects allows kubectl users to see the state of past objects (as long as they haven't been compacted by etcd) which opens the possibility for reverting accidental patches or deletions.

The ability to get stale reads from the API is only available for LIST operations but would ideally also be implemented for GET operations in a following PR.

#### Which issue(s) this PR fixes:

Fixes https://github.com/kubernetes/kubectl/issues/965

#### Does this PR introduce a user-facing change?

```release-note
Adds the --resource-version flag:
If present, sets the ResourceVersion parameter in the request
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs
kubectl get configmaps --resource-version=500 
kubectl get secrets mysecret --resource-version=60
```

### CODE REVIEW

File: staging/src/k8s.io/client-go/rest/request.go [hunk 1]
```diff
@@ -315,6 +315,18 @@ func (r *Request) NamespaceIfScoped(namespace string, scoped bool) *Request {
...
+// ResourceVersion sets the requested version of the resource as a request parameter
+func (r *Request) ResourceVersion(resourceVersion string) *Request {
+	if r.err != nil {
+		return r
+	}
+	if r.params == nil {
+		r.params = make(url.Values)
+	}
+	r.params.Set("resourceVersion", resourceVersion)
+	return r
```
Comment (apelisse) [inline staging/src/k8s.io/client-go/rest/request.go:327]:
```suggestion
	return r.setParam("resourceVersion", resourceVersion)
```

File: staging/src/k8s.io/kubectl/pkg/cmd/get/get.go [hunk 2]
```diff
@@ -182,6 +183,7 @@ func NewCmdGet(parent string, f cmdutil.Factory, streams genericiooptions.IOStre
 	cmd.Flags().BoolVar(&o.IgnoreNotFound, "ignore-not-found", o.IgnoreNotFound, "If the requested object does not exist the command will return exit code 0.")
 	cmd.Flags().StringVar(&o.FieldSelector, "field-selector", o.FieldSelector, "Selector (field query) to filter on, supports '=', '==', and '!='.(e.g. --field-selector key1=value1,key2=value2). The server only supports a limited number of field queries per type.")
 	cmd.Flags().BoolVarP(&o.AllNamespaces, "all-namespaces", "A", o.AllNamespaces, "If present, list the requested object(s) across all namespaces. Namespace in current context is ignored even if specified with --namespace.")
+	cmd.Flags().StringVar(&o.ResourceVersion, "resource-version", o.ResourceVersion, "If present, get the exact resourceVersion of the specified resource")
```
Comment (apelisse) [inline staging/src/k8s.io/kubectl/pkg/cmd/get/get.go:186]: I didn't quite follow the rest of the interaction on the bug and I don't know enough about the watch-cache, but if there are interactions between flags (resourceVersion and chunksize), then we should probably:
1. Document these in the usage
2. Check their incompatibility and return appropriate errors.

**Review by apelisse — COMMENTED** [review]
I think this could also use a "test-cmd" test to make sure everything is wired properly.

**Comment by logicalhan** [comment]
> /remove-sig api-machinery

Why remove api-machinery, client-go belongs to api-machinery

/sig api-machinery

File: staging/src/k8s.io/cli-runtime/pkg/resource/helper.go [hunk 3]
```diff
@@ -96,10 +98,18 @@ func (m *Helper) WithSubresource(subresource string) *Helper {
...
+func (m *Helper) WithResourceVersion(resourceVersion string) *Helper {
+	m.ResourceVersion = resourceVersion
+	return m
+}
+
 func (m *Helper) Get(namespace, name string) (runtime.Object, error) {
 	req := m.RESTClient.Get().
 		NamespaceIfScoped(namespace, m.NamespaceScoped).
 		Resource(m.Resource).
+		ResourceVersion(m.ResourceVersion).
```
[the remark this answers is not in the document]
Comment (logicalhan) [reply staging/src/k8s.io/cli-runtime/pkg/resource/helper.go:112]: List has different semantics. One very confusing thing about doing this is that the parameter "resource_version" is going to have different meanings when applied to GET as opposed to LIST, where if you specify resource_version, it means "no older than" whereas  in this implementation, it will mean "exactly this RV".

[the remark this answers is not in the document]
Comment (logicalhan) [reply staging/src/k8s.io/cli-runtime/pkg/resource/helper.go:112]: /cc @jpbetz @deads2k on the inconsistent semantics between "resource_version" for GET and LIST requests.

I'd prefer using "exact_resource_version" to avoid the inconsistent semantics.

**Comment by logicalhan** [comment]
This doesn't do what you think it's doing. If you're plumbing `--resource-version` to etcd, then you're actually doing a bounded stale read (i.e. equal to or greater than RV). See https://kubernetes.io/docs/reference/using-api/api-concepts/#semantics-for-get-and-list for more details.

/hold

**Review by aofekiko — COMMENTED** [review]
Apologies it took me a while to respond, I forgot to submit my comments

**Comment by aofekiko** [comment]
> This doesn't do what you think it's doing. If you're plumbing `--resource-version` to etcd, then you're actually doing a bounded stale read (i.e. equal to or greater than RV). See https://kubernetes.io/docs/reference/using-api/api-concepts/#semantics-for-get-and-list for more details.

I'd like to clarify some things.
To achieve stale reads there'd also be a need to change how the kube-apiserver works.
While I did make some code changes to make sure that plumbing a resourceVersion flag straight to etcd would work, I did not want to start a PR on such changes without an agreement on how this should be implemented.
Since I got no responses on the issue related to this PR I thought I could send this PR while the topic of allowing stale reads in the API is being discussed, as adding the flag to kubelet shouldn't be blocked by the API changes.
I apologize if I failed to convey all of that the first time.

**Comment by logicalhan** [comment]
> While I did make some code changes to make sure that plumbing a resourceVersion flag straight to etcd would work, I did not want to start a PR on such changes without an agreement on how this should be implemented. 

Yes but the PR description says: `Implements the --resource-version flag which enables stale reads at exact timestamps with kubectl by setting the ResourceVersion parameter for get and list requests.` which is factually incorrect.

**Comment by aofekiko** [comment]
> Yes but the PR description says: `Implements the --resource-version flag which enables stale reads at exact timestamps with kubectl by setting the ResourceVersion parameter for get and list requests.` which is factually incorrect.

I have updated the description to better convey the PR.
Sorry again 🙏

**Comment by jpbetz** [comment]
`resourceVersion` has a lot of historical baggage (for example, the [grid of list semantics](https://kubernetes.io/docs/reference/using-api/api-concepts/#resource-versions)) that complicate this.

For list, we don't recommend setting `resourceVersion` without also setting `resourceVersionMatch` to either `Exact` or `NotOlderThan` because of inconsistencies of how `resourceVersion=0` are handled without `resourceVersionMatch`.

get's `resourceVersion` is only capable of `NotOlderThan` semantics.

I don't think we should expose a `--resource-version` flag that doesn't address this.  A flag that allows the user to provide a `NotOlderThan` limit would be possible today, I don't know if that is useful.

**Comment by aofekiko** [comment]
Added a `--resource-version-flag` to address @jpbetz's comment.

**Comment by logicalhan** [comment]
> Added a `--resource-version-match` to address @jpbetz's comment.

I would only expose that internally, externally, I would use a flag `--exact-resource-version`. I don't imagine that bounded stale reads are very useful for GET requests.

**Comment by aofekiko** [comment]
> I would only expose that internally, externally, I would use a flag `--exact-resource-version`. I don't imagine that bounded stale reads are very useful for GET requests.

I agree, I've done as such.
Also implemented the flag for GET as well (Of course it is still lacking the implementation in the API server)

**Comment by jpbetz** [comment]
> Also implemented the flag for GET as well (Of course it is still lacking the implementation in the API server)

I don't think we should implement a the flag for GET on the client when it is not supported on the server.

**Comment by aofekiko** [comment]
> I don't think we should implement a the flag for GET on the client when it is not supported on the server.

I am planning on implementing support for stale GETs on the server as well.
Would it be better to make this PR for LIST only and then create another PR that implements stale GETs on the server or merge the two into this PR?

**Comment by aofekiko** [comment]
I have made too much of a mess on this PR and made it hard to follow.
I will be closing this PR and will create a new one once it's ready.
